### DISCUSSION TITLE
Managing fleets of datacenter servers?

### DISCUSSION THREAD

**Discussion by rusholme** [fwupd/fwupd#3077] [Q&A] [answered]
Hi,

Is it possible to use `fwupd` to manage firmware for (heterogeneous) fleets of datacenter servers (not featured in LVFS)?

Basically I would like to curate a _private_ repo of vendor firmware, then rollout updates on command across the fleet. This will allow staged deployment of updates for testing and to keep servers on pinned firmware versions.

There are specific vendor tools to do this, but AFAIK no widespread universal solution (possibly ansible/Redfish?)

I do see documentation on maintaining [offline firmware sources](https://lvfs.readthedocs.io/en/latest/offline.html) but can `fwupd` even support enterprise server updates?

Is this use case possible, or ever been considered? How much work would be involved?

Thanks!

**Comment by hughsie** [comment, marked as answer]
On Sun, 28 Mar 2021 at 20:15, rusholme ***@***.***> wrote:
> Is it possible to use fwupd to manage firmware for (heterogeneous) fleets of datacenter servers

Yes, there are several big companies doing exactly that right now. The
idea is that LVFS would provide the metadata and firmware, fwupd would
provide the mechanism to actually deploy the update, and
$something_else$ would perform the fleet management role. I pondered
writing the agent functionality (and there's even a semi-obsolete
fwupd branch for it) but the amount of overlap of
registering/deregistering/checkin/policy with other management
frameworks wasn't really worth the effort. There are all kinds of
per-site rules for this kind of thing, e.g. 10% of servers in staging
on the first day, the other 90% of staging on the next, 50% prod the
next, all by the 4th etc.

> Basically I would like to curate a private repo of vendor firmware, then rollout updates on command across the fleet.

Yes, this is sane. The
https://gitlab.com/fwupd/lvfs-website/-/blob/master/local.py script
allows you to create the metadata for a directory of .cab files, and
then you need to sign this metadata using jcat if you've enabled
signing on the client side (which you probably should do). There are
some instructions here
https://github.com/hughsie/libjcat#public-key-signing and you'll also
need to distribute the public half of the PKCS#7 certificate on all
the clients that are going to consume this.

If you'd like me to write up some more details on how to do this (and
put in the LVFS docs?) just say; it's a question I've been asked
before.

> There are specific vendor tools to do this, but AFAIK no widespread universal solution (possibly ansible/Redfish?)

Different companies do this different ways. One uses a wrapper around
fwupdagent written in Go, another uses like you said, Ansible. Redfish
isn't so useful for provisioning IMHO, I think it's more useful as an
update protocol from an fwupd point of view.

> Is this use case possible, or ever been considered? How much work would be involved?

Of course, there is another solution that at least one company is
using; you use the LVFS metadata (or a simple PULP mirror of it) and
you put the firmwares you want in an allow-list. The "fwupdmgr
set-approved-firmware" command sets the checksums of the "approved"
firmware and other firmwares are explicitly blocked from the GUI
clients and CLI commands like fwupdmgr. The advantage there is you can
use the existing LVFS signing key (as the metadata is unmodified) at
the expense of slightly larger metadata including all the devices you
have not approved.

Richard

**Reply by rusholme** [reply]
Thanks for the detailed reply!

- It would be extremely helpful to have a guide to private repos in the LVFS docs.
- I agree fleet management is better left to dedicated tools, that shouldn't be `fwupd`'s job.
- To be explicit, I'm thinking of rack-mounted servers, typically managed via IPMI, which are noticeably missing from LVFS (for example Dell have not included their PowerEdge line?) Is there any technical or security reasons such enterprise machines would not work with `fwupd`?

**Reply by superm1** [reply]
>To be explicit, I'm thinking of rack-mounted servers, typically managed via IPMI, which are noticeably missing from LVFS (for example Dell have not included their PowerEdge line?) Is there any technical or security reasons such enterprise machines would not work with fwupd?

I can't speak on behalf of Dell here specifically, but I will say that some vendors firmware update strategy doesn't involve publishing firmware updates to third parties.  Often the BMC may expect to be orchestrating more of the updates and/or have ways to apply them out of band.

**Comment by hughsie** [comment]
On Mon, 29 Mar 2021 at 17:39, rusholme ***@***.***> wrote:
> It would be extremely helpful to have a guide to private repos in the LVFS docs.

Lets start with
https://gitlab.com/fwupd/lvfs-website/-/merge_requests/924/diffs --
comments there very welcome.

Richard
